File: 22.08/21.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    if list1 == nil {
        return list2
    }
    
    if list2 == nil {
        return list1
    }
    
    if list1.Val > list2.Val {
        list1, list2 = list2, list1
    }
    
    ans := list1
    
    for list2 != nil {
        if list2.Val >= list1.Val {
            if list1.Next != nil && list1.Next.Val < list2.Val {
                list1 = list1.Next
                continue
            }
            tmp := list2.Next
            list1.Next, list2.Next = list2, list1.Next
            list1 = list1.Next
            list2 = tmp
        } else {
            if list1.Next != nil {
                list1.Next = list2
                break
            } else {
                list1 = list1.Next
            }
        }
    }
    
    return ans
}

File: typescript/23.08/21.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
    let head = null, before
    while (list1 || list2) {
        let node
        if (list1 && list2) {
            if (list1.val > list2.val) {
                node = new ListNode(list2.val)
                list2 = list2.next
            } else {
                node = new ListNode(list1.val)
                list1 = list1.next
            }
        } else if (list1) {
            node = new ListNode(list1.val)
            list1 = list1.next
        } else {
            node = new ListNode(list2.val)
            list2 = list2.next
        }
        if (!head) {
            head = node
        }
        if (before) {
            before.next = node
        }
        before = node
    }
    return head
};
